internal/http/connection/connection.go

```diff
@@ -0,0 +1,182 @@
+	return c.Conn.Close()
+}
+
+func (c *HttpConnection) SendRequest(request []byte) error {
+	if c.Callbacks.BeforeSendRequest != nil {
+		c.Callbacks.BeforeSendRequest(string(request))
+	}
+
+	return c.SendBytes(request)
+}
+
+func (c *HttpConnection) SendBytes(bytes []byte) error {
+	if c.Callbacks.BeforeSendBytes != nil {
+		c.Callbacks.BeforeSendBytes(bytes)
+	}
+
+	n, err := c.Conn.Write(bytes)
+	if err != nil {
+		return err
+	}
+
+	if n != len(bytes) {
+		return errors.New("failed to write entire bytes to connection")
+	}
+
+	return nil
+}
+
+func (c *HttpConnection) ReadResponse() (http_parser.HTTPResponse, error) {
+	return c.ReadResponseWithTimeout(2 * time.Second)
+}
+
+func (c *HttpConnection) ReadIntoBuffer() error {
+	// We don't want to block indefinitely, so we'll set a read deadline
+	c.Conn.SetReadDeadline(time.Now().Add(1 * time.Millisecond))
+
+	buf := make([]byte, 1024)
+	n, err := c.Conn.Read(buf)
+
+	if n > 0 {
+		c.UnreadBuffer.Write(buf[:n])
+	}
+
+	return err
+}
+
+func (c *HttpConnection) ReadResponseWithTimeout(timeout time.Duration) (http_parser.HTTPResponse, error) {
+	shouldStopReadingIntoBuffer := func(buf []byte) bool {
+		_, _, err := http_parser.Parse(buf)
+
+		return err == nil
+	}
+
+	c.readIntoBufferUntil(shouldStopReadingIntoBuffer, timeout)
+
+	response, readBytesCount, err := http_parser.Parse(c.UnreadBuffer.Bytes())
+
+	if c.Callbacks.AfterBytesReceived != nil && readBytesCount > 0 {
+		c.Callbacks.AfterBytesReceived(c.UnreadBuffer.Bytes()[:readBytesCount])
+	}
+
+	if err != nil {
+		return http_parser.HTTPResponse{}, err
+	}
+
+	// We've read a response! Let's remove the bytes we've read from the buffer
+	c.UnreadBuffer = *bytes.NewBuffer(c.UnreadBuffer.Bytes()[readBytesCount:])
+
+	if c.Callbacks.AfterReadResponse != nil {
+		c.Callbacks.AfterReadResponse(response)
+	}
+
+	return response, nil
+}
+
+func (c *HttpConnection) readIntoBufferUntil(condition func([]byte) bool, timeout time.Duration) {
+	deadline := time.Now().Add(timeout)
+
+	for !time.Now().After(deadline) {
+		// We'll swallow these errors and try reading again anyway
+		_ = c.ReadIntoBuffer()
+
+		if condition(c.UnreadBuffer.Bytes()) {
+			break
+		} else {
+			time.Sleep(10 * time.Millisecond) // Let's wait a bit before trying again
+		}
+	}
+}
+
+func newConn(address string) (net.Conn, error) {
+	attempts := 0
+
+	for {
+		var err error
+		var conn net.Conn
+
+		conn, err = net.Dial("tcp", address)
+
+		if err == nil {
+			return conn, nil
+		}
+
+		// Already a timeout
+		if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
+			return nil, err
+		}
+
+		// 50 * 100ms = 5s
+		if attempts > 50 {
+			return nil, err
+		}
+
+		attempts += 1
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+func (c *HttpConnection) EnsureNoUnreadData() {
+	c.ReadIntoBuffer() // Make sure there is no unread data
+
+	if c.UnreadBuffer.Len() > 0 {
+		panic(fmt.Sprintf("Found extra data: %q", c.UnreadBuffer.String()))
```

Comment:
@ryan-gang how does a panic help here? This is a user-error, so we should be sending an error message instead